python/ml4ir/applications/ranking/tests/test_ml4ir_sanity.py

```diff
@@ -0,0 +1,266 @@
+from ml4ir.applications.ranking.config.parse_args import get_args
+import shutil
+import pathlib
+import random
+from argparse import Namespace
+warnings.filterwarnings("ignore")
+
+
+def create_ml4ir_dataframe(X, Y):
+    """
+    Takes the generated data (X: data features, Y: clicked) and formulate a dataframe consumable by ml4ir
+    """
+    def add_features_to_record(X, xid, r, label):
+        for i in range(len(X[xid])):
+            r['f' + str(i)] = X[xid][i]
+        r['clicked'] = label
+
+    clicked_docs = X[Y == 1]
+    nonclicked_docs = X[Y == 0]
+    consumed_clicked_docs = 0
+    consumed_nonclicked_docs = 0
+    rows = []
+    query_count = 0
+    document_count = 0
+    while consumed_clicked_docs < len(clicked_docs) and consumed_nonclicked_docs < len(nonclicked_docs):
+        qid = 'query'+str(query_count)
+        query_count += 1
+        r = {'query_id':qid}
+        docs_count = random.randint(2, 25)
+        r['record_id'] = 'record'+str(document_count)
+        r['fr'] =docs_count
+        add_features_to_record(clicked_docs, consumed_clicked_docs, r, 1)
+        consumed_clicked_docs += 1
+        document_count += 1
+        rows.append(r)
+        for i in range(docs_count-1):
+            if consumed_nonclicked_docs < len(nonclicked_docs):
+                r = {'query_id': qid}
+                r['fr'] = i+1
+                r['record_id'] = 'record' + str(document_count)
+                add_features_to_record(nonclicked_docs, consumed_nonclicked_docs, r, 0)
+                consumed_nonclicked_docs += 1
+                document_count += 1
+                rows.append(r)
+            else:
+                break
+    return pd.DataFrame(rows)
+
+
+def generate_data(args):
+    """
+    Generate data with the specified parameters
+    - n_classes: Number of classes in the datasets. Should be 2 (clicked, not clicked)
+    - n_samples: Number of samples to generate
+    - n_features: Number of features per sample
+    - n_redundant: The number of redundant features
+    - n_informative: The number of informative (Non-redundant) features
+    - n_clusters_per_class: The number of clusters per class.
+    - flip_y: The fraction of samples whose class is assigned randomly. Larger
+        values introduce noise in the labels and make the classification
+        task harder.
+    - class_sep: Larger values spread out the clusters/classes and make the classification task easier.
+    - seed: Random seed
+    - n_trials: Number of trials of data generation to obtain a linearly separable data
+    Returns: the best seprabale data set by a Perceptorn.
+    """
+    separable = False
+    best_accuracy = 0
+    best_accuracy_data = None
+    random.seed(args.seed)
+    seeds = list(range(args.n_trials))
+    random.shuffle(seeds)
+    trials = len(seeds)-1
+    while not separable and trials > 0:
+        samples = datasets.make_classification(n_classes=args.n_classes, n_samples=args.n_samples, n_features=args.n_features,
```

Comment:
@mohazahran Do we need to generate the data multiple times to get linearly separable? Does setting `class_sep` to a really high value not work out of the box? I'm just trying to understand the reasoning behind the 100 trials here.